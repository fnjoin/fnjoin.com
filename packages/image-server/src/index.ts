import path from "path";
import express from "express";
import { glob } from "glob"; // Install with `npm install @types/glob glob`
import sharp from "sharp";
import yargs from "yargs";
import { hideBin } from "yargs/helpers";

async function main() {
    const argv = await yargs(hideBin(process.argv))
        .option("root-dir", {
            type: "string",
            description: "Root directory for images",
            demandOption: true, // Make the root argument required
        })
        .option("port", {
            type: "number",
            description: "Port to listen on",
            default: 3000,
        })
        .option("quality", {
            type: "number",
            description: "Quality of webp images",
            default: 85,
        })
        .option("default-width", {
            type: "number",
            description: "Default width for images",
            default: 800,
        }).argv;

    // Extract the root folder path from the startup arguments
    const rootFolder: string = argv.rootDir;

    const app: express.Application = express();
    const port: number = argv.port;

    app.get("/*", async (req, res) => {
        const { w, q } = req.query;
        const width: number = parseInt(w as string, 10) || argv.defaultWidth;
        const quality: number = parseInt(q as string, 10) || argv.quality;

        const requestedPath: string = path.join(rootFolder, req.path);
        const baseName: string = path.basename(
            requestedPath,
            path.extname(requestedPath),
        );

        // Use glob to find any file with the same basename but different extension
        const files = await glob(
            `${path.dirname(requestedPath)}/${baseName}.*`,
        );

        if (!files.length) {
            console.log("File not found:", requestedPath);
            console.log("Base name:", baseName);
            console.log("Files:", files);
            return res.status(404).send("File not found");
        }

        // Assume the first match is the correct file
        const originalImagePath: string = files[0];

        return sharp(originalImagePath)
            .resize(width)
            .webp({ quality })
            .toBuffer()
            .then((data) => res.type("webp").send(data))
            .catch((sharperr) => res.status(500).send(sharperr.message));
    });

    app.listen(port, () => {
        console.log(`Server running on port ${port}`);
    });
}

main()
    .then(() => console.log("done"))
    .catch((err) => console.error("error", err));
