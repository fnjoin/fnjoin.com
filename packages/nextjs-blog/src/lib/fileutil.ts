import * as fs from "fs";
import * as path from "path";

export interface CrawlInput {
    dir: string;
    filter?: (file: string) => boolean;
}

export function* crawlDirectories({
    dir,
    filter,
}: CrawlInput): Generator<string, void, string[]> {
    for (let file of fs.readdirSync(dir)) {
        let p = path.join(dir, file);
        if (fs.statSync(p).isDirectory()) {
            yield* crawlDirectories({ dir: p, filter });
        } else if (filter && filter(p)) {
            yield p;
        } else if (!filter) {
            yield p;
        }
    }
}
