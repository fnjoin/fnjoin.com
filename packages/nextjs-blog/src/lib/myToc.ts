import toc from "mdast-util-toc";

export interface TocOptions {
    tight?: boolean;
    heading?: string;
}

export function extractToc(options: TocOptions) {
    const settings = {
        ...options,
        tight:
            options && typeof options.tight === "boolean"
                ? options.tight
                : true,
    };

    return function (tree: any): void {
        const result = toc(tree, settings);
        if (
            result.endIndex === undefined ||
            result.endIndex === -1 ||
            result.index === undefined ||
            result.index === -1 ||
            !result.map
        ) {
            return;
        }

        tree.children = [result.map];
    };
}
