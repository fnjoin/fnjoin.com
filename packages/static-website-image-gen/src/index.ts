import fs from "fs";
import path from "path";
import { glob } from "glob";
import { JSDOM } from "jsdom";
import sharp, { FormatEnum } from "sharp";
import yargs from "yargs";
import { hideBin } from "yargs/helpers";

// Function to process each HTML file
async function processHtmlFiles(
    htmlDirectory: string,
    originDirectory: string,
    outputDirectory: string,
    websiteBasePath?: string,
) {
    // Find all HTML files in the given directory
    const htmlFiles = await glob.glob(`${htmlDirectory}/**/*.html`);

    // Process each HTML file
    for (const file of htmlFiles) {
        console.log("found file", file);
        const htmlContent = fs.readFileSync(file, "utf-8");
        const dom = new JSDOM(htmlContent);
        const images = dom.window.document.querySelectorAll("img");

        // Process each img tag
        for (const img of images) {
            console.log("found img", img);
            const src = img.getAttribute("src");
            const srcset = img.getAttribute("srcset");

            const imageSources = new Set<string>();

            if (src) imageSources.add(src);
            if (srcset) {
                srcset.split(",").forEach((part) => {
                    const srcInfo = part.trim().split(" ")[0];
                    imageSources.add(srcInfo);
                });
            }

            // Process each image source
            for (const source of imageSources) {
                await processImageSource(
                    source,
                    originDirectory,
                    outputDirectory,
                    websiteBasePath,
                );
            }
        }
    }
}

// Function to process each image source
async function processImageSource(
    source: string,
    originDirectory: string,
    outputDirectory: string,
    websiteBasePath?: string,
) {
    console.log("processing image source", source);
    const { dir, name } = path.parse(source);
    const match = name.match(/(.*?)\.w(\d+)q(\d+)$/);
    if (!match) {
        console.warn(`Invalid image name: ${name}`);
        console.warn(`Skipping image ${source}`);
        return;
    }

    // get ext from source
    const format = path.extname(source).slice(1);
    console.log("format", format);
    const [, baseName, width, quality] = match;
    const originPath = path.join(
        originDirectory,
        websiteBasePath && websiteBasePath.startsWith("/")
            ? dir.replace(websiteBasePath, "")
            : dir,
        baseName,
    );
    const outputPath = path.join(
        outputDirectory,
        websiteBasePath && websiteBasePath.startsWith("/")
            ? source.replace(websiteBasePath, "")
            : source,
    );

    // Find the original file with extension jpg, png, etc.
    const originalFile = glob.sync(`${originPath}.{jpg,jpeg,png,webp}`)[0];
    if (!originalFile) {
        console.warn(
            "Broken image!",
            `original file not found for ${source} at ${originPath}.{jpg,jpeg,png,webp}`,
            "origin dir",
            originDirectory,
            "baseName",
            baseName,
            "dir",
            dir,
        );
        return;
    }
    console.log("origin file found", originalFile);

    // return if the output file exists already
    if (fs.existsSync(outputPath)) {
        // console.log("output file exists", outputPath);
        return;
    }

    // Ensure the output directory exists
    fs.mkdirSync(path.dirname(outputPath), { recursive: true });

    // Convert and save the image
    await sharp(originalFile)
        .resize(parseInt(width), null)
        .toFormat(format as keyof FormatEnum, { quality: parseInt(quality) })
        .toFile(outputPath);
}

export default async function main(): Promise<void> {
    // Example usage

    const argv = await yargs(hideBin(process.argv))
        .option("html-dir", {
            type: "string",
            description: "HTML files directory",
            demandOption: true,
        })
        .option("origin-dir", {
            type: "string",
            description: "Root directory for images",
            demandOption: true, // Make the root argument required
        })
        .option("target-dir", {
            type: "string",
            description: "Target directory for images",
            demandOption: true,
        })
        .option("website-base-path", {
            type: "string",
            description: "If your website is not on /, add your basePath here",
        })
        .option("quality", {
            type: "number",
            description: "Quality of webp images",
            default: 85,
        })
        .option("default-width", {
            type: "number",
            description: "Default width for images",
            default: 800,
        }).argv;

    processHtmlFiles(
        argv.htmlDir,
        argv.originDir,
        argv.targetDir,
        argv.websiteBasePath,
    )
        .then(() => console.log("Image processing completed."))
        .catch((err) => console.error(err));
}
