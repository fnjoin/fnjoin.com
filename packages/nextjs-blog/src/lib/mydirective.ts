import { h } from "hastscript";
import { Element } from "hastscript/lib/create-h";
import { Root } from "remark-parse/lib";

import { Node } from "unified/lib";
import { visit } from "unist-util-visit";

// This plugin is an example to let users write HTML with directives.
// Itâ€™s informative but rather useless.
// See below for others examples.
interface DirectiveNode extends Node {
    type: "containerDirective" | "leafDirective" | "textDirective";
    name: string;
    data?: { hName?: string; hProperties?: Record<string, unknown> };
    attributes?: Record<string, any> | null;
}

interface ReferenceCounter {
    [type: string]: {
        [ref: string]: number;
    };
}
class ReferenceCounterByType {
    refs: ReferenceCounter = {};
    counters: { [type: string]: number } = {};
    constructor() {}
    get(type: string, ref: string | null | undefined) {
        if (!ref) {
            return null;
        }
        if (!this.counters[type]) {
            this.counters[type] = 1;
        }
        if (!this.refs[type]) {
            this.refs[type] = {};
        }
        if (!this.refs[type][ref]) {
            this.refs[type][ref] = this.counters[type]++;
        }
        return this.refs[type][ref];
    }
}

export function myRemarkPlugin() {
    let sections = 0;

    return function transformer(tree: Root): void {
        // Correctly specify each type to visit in separate calls or a workaround
        const typesToVisit: DirectiveNode["type"][] = [
            "containerDirective",
            "leafDirective",
            "textDirective",
        ];
        const refsByType = new ReferenceCounterByType();

        typesToVisit.forEach((type) => {
            visit<Root, typeof type>(tree, type, (node: DirectiveNode) => {
                const data = node.data || (node.data = {});
                const hast: Element = h(node.name, node.attributes || {});
                // data.hName = hast.tagName;
                data.hProperties = hast.properties;
                data.hProperties.dtype = type;

                let ns = "global";
                if (
                    hast.properties.id &&
                    hast.properties.id.toString().indexOf(":") > -1
                ) {
                    ns = hast.properties.id.toString().split(":")[0];
                }
                data.hProperties.refns = ns;

                if (hast.tagName === "ref") {
                    // console.log("ref", data, hast);
                    const refnum = refsByType.get(
                        ns,
                        hast.properties.id as string,
                    );
                    data.hProperties["data-element"] = "ref";
                    data.hName = "span";
                    data.hProperties.refnum = refnum;
                    // console.log("refnum", refnum);
                }
                if (hast.tagName === "full-bleed") {
                    data.hProperties["data-element"] = "full-bleed";
                    data.hName = "div";
                }
                if (hast.tagName === "inline-callout") {
                    // console.log(hast, data);
                    data.hProperties["data-element"] = "inline-callout";
                    data.hName = "div";
                }
                if (hast.tagName === "callout") {
                    data.hProperties["data-element"] = "callout";
                    data.hName = "div";
                }
                if (hast.tagName === "margin-note") {
                    data.hProperties["data-element"] = "margin-note";
                    data.hName = "div";
                }
                if (hast.tagName === "figure-fence") {
                    data.hProperties["data-element"] = "figure-fence";
                    data.hProperties.refnum = refsByType.get(
                        ns,
                        hast.properties.id as string,
                    );
                    data.hName = "div";
                }
                if (type === "containerDirective") {
                    sections++;
                    data.hProperties.sections = sections;
                }
                // console.log("properties", hast.properties, data);
            });
        });
    };
}
