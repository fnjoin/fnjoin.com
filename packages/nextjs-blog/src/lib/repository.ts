import * as fs from "fs";
import * as path from "path";
import matter from "gray-matter";
import { crawlDirectories } from "./fileutil";
import { AuthorProps, MyPost } from "@/interfaces/mypost";

function getAuthorMarkdown(fullPath: string): AuthorProps {
    const fileContents = fs.readFileSync(fullPath, "utf8");
    const { data } = matter(fileContents);
    return data as AuthorProps;
}

function getMarkdown(fullPath: string): MyPost {
    const fileContents = fs.readFileSync(fullPath, "utf8");
    const { data, content } = matter(fileContents);
    return { ...data, content } as MyPost;
}

export interface BlogRepositoryProps {
    dir: string;
}
export class BlogRepository {
    static fromCwd() {
        return new BlogRepository({
            dir: path.join(process.cwd(), "../../content"),
        });
    }
    private dir: string;
    private posts: MyPost[] = [];
    private postsByPath: Record<string, MyPost> = {};

    constructor({ dir }: BlogRepositoryProps) {
        this.dir = dir;
        const files = crawlDirectories({
            dir: this.dir,
            filter: (file) =>
                file.endsWith(".md") &&
                file.indexOf("/authors/") < 0 &&
                file.indexOf("post") > 0,
        });
        let x: IteratorResult<string> = files.next();
        while (!x.done) {
            const post = getMarkdown(x.value);
            if (post.author === "Salman Malik") {
                post.author_detail = getAuthorMarkdown(
                    path.join(dir, "authors/smalik.md"),
                );
            } else if (post.author === "Archie Cowan") {
                post.author_detail = getAuthorMarkdown(
                    path.join(dir, "authors/acowan.md"),
                );
            } else if (fs.existsSync(`authors/${post.author}.md`)) {
                post.author_detail = getAuthorMarkdown(
                    path.join(dir, `authors/${post.author}.md`),
                );
            }

            if (!post.coverImage) {
                // post.coverImage = {
                //     imageSrc: "/assets/astronaut.jpg",
                //     caption: "Placeholder image of an astronaut.",
                // };
            }

            post.slug = x.value.replace(this.dir + "/", "").replace(".md", "");
            // console.log("post slug:", post.slug);
            this.posts.push(post);
            this.postsByPath[post.slug] = post;
            x = files.next();
        }

        // sort posts in reverse chronological order
        this.posts.sort((a, b) => {
            return new Date(b.date).getTime() - new Date(a.date).getTime();
        });
    }

    getAllPosts(): MyPost[] {
        return this.posts;
    }

    getPostByPath(slug: string): MyPost {
        return this.postsByPath[slug];
    }
}

export class PageRepository {
    static fromCwd() {
        return new PageRepository({
            dir: path.join(process.cwd(), "../../content"),
        });
    }
    private dir: string;
    private posts: MyPost[] = [];
    private postsByPath: Record<string, MyPost> = {};

    constructor({ dir }: BlogRepositoryProps) {
        this.dir = dir;
        const files = crawlDirectories({
            dir: this.dir,
            filter: (file) =>
                file.endsWith(".md") &&
                file.indexOf("/authors/") < 0 &&
                file.indexOf("/page/") > 0,
        });
        let x: IteratorResult<string> = files.next();
        while (!x.done) {
            const post = getMarkdown(x.value);

            post.slug = x.value.replace(this.dir + "/", "").replace(".md", "");
            // console.log("post slug:", post.slug);
            this.posts.push(post);
            this.postsByPath[post.slug] = post;
            x = files.next();
        }
    }
    getAllPosts(): MyPost[] {
        return this.posts;
    }

    getPostByPath(slug: string): MyPost {
        return this.postsByPath[slug];
    }
}
