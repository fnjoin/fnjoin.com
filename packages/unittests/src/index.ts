import { h } from "hastscript";
import { Element } from "hastscript/lib/create-h";
import { Root } from "remark-parse/lib";

import { Node } from "unified/lib";
import { visit } from "unist-util-visit";

export class Hello {
    public sayHello() {
        return "hello, world!";
    }
}

// This plugin is an example to let users write HTML with directives.
// Itâ€™s informative but rather useless.
// See below for others examples.
interface DirectiveNode extends Node {
    type: "containerDirective" | "leafDirective" | "textDirective";
    name: string;
    data?: { hName?: string; hProperties?: Record<string, unknown> };
    attributes?: Record<string, any> | null;
}

export function myRemarkPlugin() {
    let sections = 0;

    return function transformer(tree: Root): void {
        // Correctly specify each type to visit in separate calls or a workaround
        const typesToVisit: DirectiveNode["type"][] = [
            "containerDirective",
            "leafDirective",
            "textDirective",
        ];

        typesToVisit.forEach((type) => {
            visit<Root, typeof type>(tree, type, (node: DirectiveNode) => {
                const data = node.data || (node.data = {});
                const hast: Element = h(node.name, node.attributes || {});
                data.hName = hast.tagName;
                data.hProperties = hast.properties;
                data.hProperties.dtype = type;
                if (type === "containerDirective") {
                    sections++;
                    data.hProperties.sections = sections;
                }
            });
        });
    };
}
