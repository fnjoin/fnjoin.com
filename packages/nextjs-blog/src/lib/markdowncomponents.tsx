import * as fs from "fs";
import * as path from "path";
import matter from "gray-matter";
import imageSize from "image-size";
import Image from "next/image";
import React, { PropsWithChildren } from "react";
import ReactMarkdown from "react-markdown";
import remarkDirective from "remark-directive";
import remarkGfm from "remark-gfm";
import remarkParse from "remark-parse";
import markdownStyles from "@/app/_components/markdown-styles.module.css";
import { AuthorProps, MyPost } from "@/interfaces/mypost";
import { myRemarkPlugin } from "@/lib/mydirective";
import { extractToc } from "@/lib/myToc";

const projectRoot = process.cwd();

// Note!! Headings are set in markdown-styles.module.css
const gridStyle = ""; // "col-start-4 col-end-10 col-span-7";
const marginStyle = ""; // "md:col-start-11 md:col-end-13 md:col-span-3";
const extraWideStyle = ""; // "md:col-start-2 md:col-end-10 md:col-span-10";

export function getMarkdown(): MyPost {
    const fullPath = path.join(projectRoot, `src/app/hellomarkdown/file.md`);
    const fileContents = fs.readFileSync(fullPath, "utf8");
    const { data, content } = matter(fileContents);

    return { ...data, slug: "/hellomarkdown", content } as MyPost;
}

export function Article({ children }: PropsWithChildren) {
    return <div className={markdownStyles.article}>{children}</div>;
}

export function Paragraph({ children }: any) {
    return (
        // TODO if this a p tag, then we can't have nested divs under it, also figures aren't allowed
        // maybe we can check the children and conditionally render as a paragraph if the children are just text, does this matter?
        <div className={`m-2 ${gridStyle}`}>{children}</div>
    );
}

export function MarginNote({ children }: any) {
    return (
        <div
            className={`m-2 ${gridStyle} row-span-4 ${marginStyle} text-sm ${markdownStyles["margin-note"]}`}
        >
            {children}
        </div>
    );
}

export interface TagProps {
    tags?: string[];
}
export function Tags({ tags }: TagProps) {
    if (!tags) {
        return <></>;
    }
    return (
        <div className={`m-2 ${gridStyle}`}>
            <div className="flex gap-1 flex-wrap text-sm justify-center">
                <span>Tags:</span>
                {tags?.map((tag, idx) => (
                    <span
                        key={idx}
                        className="bg-gray-100 rounded-full px-3 font-semibold text-gray-600"
                    >
                        #{tag}
                    </span>
                ))}
            </div>
        </div>
    );
}

export function InlineCallout({ children }: any) {
    return (
        <div
            className={`col-span-12 col-start-1 col-end-13 ${extraWideStyle} place-content-center`}
        >
            <div className="m-5 text-lg">{children}</div>
        </div>
    );
}
export function Callout({ children }: any) {
    return (
        <div
            className={`m-2 col-start-1 col-end-13 col-span-12 ${marginStyle} md:row-span-12 text-2xl`}
        >
            {children}
        </div>
    );
}

export function Author(props: AuthorProps) {
    if (!props.picture || !props.name) {
        console.log("missing required props in Author");
        return <></>;
    }
    const size = imageSize(
        path.join(projectRoot, "../../content", props.picture),
    );
    // console.log("markdown styles", markdownStyles);
    return (
        <div className={markdownStyles.author}>
            <Image
                src={props.picture}
                width={size.width}
                height={size.height}
                alt={props.name}
                className="rounded-xl p-1 w-full"
            />
            <div className="flex items-center">
                <div className="m-2 text-sm">
                    <div className="prose-h4 font-semibold">
                        By {props.name}
                    </div>
                    <div>{props.bio}</div>
                </div>
            </div>
        </div>
    );
}

export function Heading1({ children }: any) {
    let id = children;
    if (typeof children === "string") {
        id = children.toLowerCase().replaceAll(" ", "-");
    }
    return <h1 id={id}>{children}</h1>;
}

export function Heading2({ children }: any) {
    // lower case, - separated
    let id = children;
    if (typeof children === "string") {
        id = children.toLowerCase().replaceAll(" ", "-");
    }
    // children.map((x: string) => x.toLowerCase().replaceAll(" ", "-"));
    return <h2 id={id}>{children}</h2>;
}

export interface ImageWithinFigureProps {
    src?: string;
    width?: number;
    height?: number;
    alt?: string;
}
export function ImageWithinFigure({
    src,
    width,
    height,
    alt,
}: ImageWithinFigureProps) {
    if (!src || !alt) {
        return <span>Missing image properties</span>;
    }
    if (!width || !height) {
        const size = imageSize(path.join(projectRoot, "../../content", src));
        width = size.width;
        height = size.height;
    }
    return (
        <Image
            src={src}
            width={width || 1024}
            height={height || 1024}
            alt={alt}
            className="rounded-xl shadow-2xl"
        />
    );
}

export interface FigureFenceProps {
    children?: any;
    caption?: string;
    title?: string;
    id?: string;
    refnum?: number;
    refns?: string;
    reflabels?: {
        [label: string]: string;
    };
}
export function FigureFence(props: FigureFenceProps) {
    let refLabel = "Figure";
    if (props.refns && props.reflabels && props.reflabels[props.refns]) {
        refLabel = props.reflabels[props.refns];
    } else {
        logMissingFigureNS();
    }

    return (
        <>
            <figure className={gridStyle} id={props.id}>
                {props.title && (
                    <div className={`${gridStyle} text-center text-md`}>
                        {props.refnum && `${props.refnum}: `}
                        {props.title}
                    </div>
                )}
                <div className="col-start-1 col-end-13 col-span-12 ">
                    {props.children}
                </div>
                <figcaption className="m-4 col-span-4 text-sm text-center">
                    {props.refnum && `${refLabel} ${props.refnum}: `}
                    {props.caption}
                </figcaption>
            </figure>
        </>
    );
}

export interface RefLinkProps {
    refid?: string;
    refnum?: number;
    reflabels?: {
        [label: string]: string;
    };
    children?: any;
}
export function RefLink(props: RefLinkProps) {
    let refLabel = "Figure";
    let refns = props.refid?.split(":")[0];
    if (refns && props.reflabels && props.reflabels[refns]) {
        refLabel = props.reflabels[refns];
    } else {
        logMissingFigureNS();
    }

    return (
        <a href={`#${props.refid}`}>
            {refLabel} {props.refnum}
        </a>
    );
}

function logMissingFigureNS() {
    console.warn(
        `
Missing ref namespaces in the article frontmatter. Set figurens in your frontmatter
e.g.:
figurens:
    img: Image
    fig: Figure
    code: Code Excerpt
        `.trim(),
    );
}

export function FullBleedImage({
    src,
    width,
    height,
    alt,
}: ImageWithinFigureProps) {
    if (!src || !alt) {
        return <span>Missing image properties</span>;
    }
    if (!width || !height) {
        // TODO maybe it would be better to get this in the markdown plugin
        const size = imageSize(path.join(projectRoot, "../../content", src));
        // TODO height width seem to be reversed in image-size???
        width = size.height;
        height = size.width;
    }
    return (
        <div className="col-start-1 col-end-13 col-span-12">
            <div className="col-start-1 col-end-13 col-span-12 ">
                <figure>
                    <Image
                        src={src}
                        alt={alt}
                        width={height}
                        height={width}
                        className="col-start-1 col-end-13 col-span-12 w-full"
                    />
                    <figcaption className="text-center">{alt}</figcaption>
                </figure>
            </div>
        </div>
    );
}

export function TocFromMarkdown({ markdown }: { markdown: string }) {
    return (
        <ReactMarkdown
            remarkPlugins={[
                remarkParse,
                remarkDirective,
                myRemarkPlugin,
                extractToc,
            ]}
        >
            {markdown}
        </ReactMarkdown>
    );
}

export function ArticleBodyFromMarkdown({ art }: { art: MyPost }) {
    return (
        <ReactMarkdown
            remarkPlugins={[
                remarkParse,
                remarkGfm,
                remarkDirective,
                myRemarkPlugin,
            ]}
            // rehypePlugins={[]}
            urlTransform={(url) => {
                return url; // or return a new url string.
            }}
            // default is all
            // allowedElements={["img", "div", "p"]}
            components={{
                span: ({ node, ...rest }) => {
                    if (node?.properties["data-element"] === "ref") {
                        return (
                            // <a href={`#${node?.properties.id}`}>
                            //     Figure {node?.properties.refnum}
                            // </a>
                            <RefLink
                                reflabels={art.figurens}
                                refid={node?.properties.id as string}
                                refnum={node?.properties.refnum as number}
                            />
                        );
                    }
                    return <span {...rest} />;
                },
                ol: ({ node, className, ...rest }) => {
                    return (
                        <ol
                            className={`${className ?? ""} ${gridStyle} list-decimal`}
                            {...rest}
                        />
                    );
                },
                ul: ({ node, className, ...rest }) => {
                    return (
                        <ul
                            className={`${className ?? ""} ${gridStyle} list-disc`}
                            {...rest}
                        />
                    );
                },
                table: ({ node, className, ...rest }) => {
                    return (
                        <table
                            className={`${className} ${gridStyle}`}
                            {...rest}
                        />
                    );
                },
                section: ({ node, className, ...rest }) => {
                    return (
                        <section
                            className={`${className} ${gridStyle}`}
                            {...rest}
                        />
                    );
                },
                div: ({ node, ...rest }) => {
                    if (node?.properties["data-element"] === "inline-callout") {
                        // console.log("page.tsx", node, rest);
                        return <InlineCallout {...rest} />;
                    }
                    if (node?.properties["data-element"] === "callout") {
                        return <Callout {...rest} />;
                    }
                    if (node?.properties["data-element"] === "figure-fence") {
                        return (
                            <FigureFence
                                reflabels={art.figurens}
                                {...rest}
                                {...node.properties}
                            />
                        );
                    }
                    if (node?.properties["data-element"] === "full-bleed") {
                        return (
                            <FullBleedImage
                                src={(node?.properties.src as string) || ""}
                                alt={(node?.properties.alt as string) || ""}
                            />
                        );
                    }
                    if (node?.properties["data-element"] === "margin-note") {
                        return <MarginNote {...rest} />;
                    }
                    return <div {...rest} />;
                },
                h2: ({ node, ...rest }) => <Heading2 {...rest} />,
                h1: ({ node, ...rest }) => <Heading1 {...rest} />,
                blockquote: ({ node, ...rest }) => {
                    return <blockquote {...rest} className={gridStyle} />;
                },
                p: ({ node, ...rest }) => {
                    return <Paragraph {...rest} />;
                },
                img: ({ node }) => {
                    return (
                        <ImageWithinFigure
                            src={node?.properties.src as string}
                            alt={node?.properties.alt as string}
                        />
                    );
                },
                pre: ({ node, ...rest }) => {
                    return (
                        <pre className={`m-2 ${gridStyle}`}>
                            {rest.children}
                        </pre>
                    );
                },
            }}
        >
            {art.content}
        </ReactMarkdown>
    );
}
