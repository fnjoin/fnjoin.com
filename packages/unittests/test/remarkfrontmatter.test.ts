import { test, expect } from "@jest/globals";
// import rehypeDocument from "rehype-document";
// import rehypeFormat from "rehype-format";
// import rehypeStringify from "rehype-stringify";
import remarkFrontmatter from "remark-frontmatter";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
// import remarkDirective from "remark-directive";

import { read } from "to-vfile";
import { matter } from "vfile-matter";

/**
 * Parse YAML frontmatter and expose it at `file.data.matter`.
 *
 * @returns
 *   Transform.
 */
export default function myUnifiedPluginHandlingYamlMatter() {
    /**
     * Transform.
     *
     * @param {Node} tree
     *   Tree.
     * @param {VFile} file
     *   File.
     * @returns {undefined}
     *   Nothing.
     */
    return function (_: any, file: VFile) {
        matter(file);
    };
}

// import { read } from "to-vfile";
import { unified } from "unified";
import { VFile } from "vfile-matter/lib";
import rehypeStringify from "rehype-stringify";
import remarkDirective from "remark-directive";
import { myRemarkPlugin } from "../src";
// import { myRemarkPlugin } from "../src/index"; // TODO why doesn't @/ work?
// import { reporter } from "vfile-reporter";

// import * as path from "path";

test("how to make frontmatter work", async () => {
    const file = await unified()
        .use(remarkParse)
        .use(rehypeStringify)
        .use(remarkFrontmatter, ["yaml"])
        .use(myUnifiedPluginHandlingYamlMatter)
        .use(remarkDirective)
        .use(myRemarkPlugin)
        .use(remarkRehype)
        .process(await read("test/remarkfrontmatter.test.md"));

    expect(file.data).toMatchSnapshot();
});
