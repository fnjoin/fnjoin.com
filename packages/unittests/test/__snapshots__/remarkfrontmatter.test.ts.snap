// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`how to make frontmatter work 1`] = `
{
  "matter": {
    "field": "value",
  },
}
`;
